### PR TITLE
ECO-3474 Adyen fixes

### PR DESCRIPTION
**Pull request by vhrychenko**
- Developer(s): @vhrychenko

- Ticket: https://spryker.atlassian.net/browse/ECO-3474

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Adyen          | minor                 |                       |

-----------------------------------------

##### Change log

Improvements

- Deprecated `AdyenControllerProvider`.
- Introduced `AdyenRouteProviderPlugin`.
- Adjusted payment redirect behavior. If payment is canceled by the user request will be redirected to the checkout payment step.

### CODE REVIEW

File: src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceBridge.php [hunk 1]
```diff
@@ -23,12 +23,12 @@ public function __construct($textService)
...
-     * @param string $algorithm
      * @param mixed $value
+     * @param string $algorithm
      *
      * @return string
      */
-    public function hashValue($algorithm, $value): string
+    public function hashValue($value, $algorithm): string
     {
         return $this->textService->hashValue($algorithm, $value);
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceBridge.php:33]: This line should changed accordingly.

File: src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceBridge.php [hunk 2]
```diff
@@ -23,12 +23,12 @@ public function __construct($textService)
...
 
     /**
-     * @param string $algorithm
      * @param mixed $value
+     * @param string $algorithm
      *
      * @return string
      */
-    public function hashValue($algorithm, $value): string
+    public function hashValue($value, $algorithm): string
```
Comment (dereuromark) [inline src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceBridge.php:31]: This looks like a dangerous (and easily overlooked) change. Is this correct?

File: src/SprykerEco/Service/Adyen/Generator/AdyenGenerator.php [hunk 3]
```diff
@@ -41,7 +41,7 @@ public function generateReference(QuoteTransfer $quoteTransfer): string
             $quoteTransfer->getTotals()->getHash(),
         ];
 
-        $string = $this->textService->hashValue(implode(static::PARAMETERS_SEPARATOR, $parameters), Hash::SHA256);
+        $string = $this->textService->hashValue(Hash::SHA256, implode(static::PARAMETERS_SEPARATOR, $parameters));
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Service/Adyen/Generator/AdyenGenerator.php:44]: Please check this as well.

File: src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeBridge.php [hunk 4]
```diff
@@ -29,7 +29,7 @@ public function __construct($salesFacade)
      *
      * @return \Generated\Shared\Transfer\OrderTransfer
      */
-    public function getOrderByIdSalesOrder(int $idSalesOrder): OrderTransfer
```
Comment (dereuromark) [inline src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeBridge.php:32]: Why removing? Is this not an int?

File: src/SprykerEco/Yves/Adyen/Plugin/Router/AdyenRouteProviderPlugin.php [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    /**
+     * @param \Spryker\Yves\Router\Route\RouteCollection $routeCollection
+     *
+     * @return \Spryker\Yves\Router\Route\RouteCollection
+     */
+    protected function addAdyenNotificationRoute(RouteCollection $routeCollection): RouteCollection
+    {
+        $routeCollection->add(
+            'adyen-notification',
+            $this->buildRoute('/adyen/notification', static::BUNDLE_NAME, static::NOTIFICATION_CONTROLLER_NAME, 'index')
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Yves/Adyen/Plugin/Router/AdyenRouteProviderPlugin.php:46]: Only post allowed here.
Probably we need to add allowed methods for each route.

File: src/SprykerEco/Zed/Adyen/Communication/Plugin/AdyenPaymentMethodFilterPlugin.php [hunk 6]
```diff
@@ -20,6 +20,7 @@
 class AdyenPaymentMethodFilterPlugin extends AbstractPlugin implements PaymentMethodFilterPluginInterface
 {
     /**
+     * {@inheritDoc}
      * Specification:
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Zed/Adyen/Communication/Plugin/AdyenPaymentMethodFilterPlugin.php:24]: Please remove this line.

File: src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeBridge.php (hunk 4 above)
Comment (aleksey-kotsuba) [reply src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeBridge.php:32]: I suppose sniffer did it. In SalesFacade it looks like
`public function getOrderByIdSalesOrder($idSalesOrder)`

File: src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceBridge.php (hunk 2 above)
Comment (vhrychenko) [reply src/SprykerEco/Service/Adyen/Dependency/Service/AdyenToUtilTextServiceBridge.php:31]: \Spryker\Service\UtilText\UtilTextServiceInterface::hashValue

File: src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeBridge.php (hunk 4 above)
Comment (dereuromark) [reply src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeBridge.php:32]: Ther sniffer does not remove typehints. if they are already in place we might just want to keep them.

Comment (vhrychenko) [reply src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeBridge.php:32]: I got a sniffer error that the bridge signature should be completely the same. That's why I've removed type hint to not have this error while committing the changes

Comment (dereuromark) [reply src/SprykerEco/Zed/Adyen/Dependency/Facade/AdyenToSalesFacadeBridge.php:32]: I see, ok.

**Pull request merged**